lib/record_store/provider.rb

```diff
@@ -128,6 +128,36 @@ def master_nameserver_for(zone_name)
 
         dns.getresource(zone_name, Resolv::DNS::Resource::IN::SOA).mname.to_s
       end
+
+      def retry_on_connection_errors(
+        max_timeouts: 5,
+        max_conn_resets: 5,
+        delay: 1,
+        backoff_multiplier: 2,
+        max_backoff: 10
+      )
+        loop do
+          begin
+            return yield
+          rescue Net::OpenTimeout
+            raise if max_timeouts <= 0
+            max_timeouts -= 1
+
+            $stderr.puts("Retrying after a connection timeout")
+          rescue Errno::ECONNRESET
+            raise if max_conn_resets <= 0
+            max_conn_resets -= 1
+
+            $stderr.puts("Retrying in #{delay}s after a connection reset")
+            backoff_sleep(delay)
+            delay = [delay * backoff_multiplier, max_backoff].min
+          end
+        end
+      end
+
+      def backoff_sleep(delay)
+        sleep(delay)
+      end
```

Comment:
Would be nice to be able to reuse the logic of https://github.com/Shopify/record_store/blob/ea938f03380e1d42f5622cabf4bf5650ca63e097/lib/record_store/provider/provider_utils/rate_limit.rb (maybe by creating a parent class `Waiter` for instance, and just having the `sleep_for` being inherited with a different message, so we could in this case add `ConnectionResetWaiter` that would show the `waiting #{delay}s to retry after a connection reset` message ?)

Reply:
Def something for a future PR for sure.